packages/agent/src/canisterStatus/index.ts

```diff
@@ -0,0 +1,242 @@
+ */
+export interface MetaData {
+  kind: 'medadata';
+  key: string;
+  path: string | ArrayBuffer;
+  decodeStrategy: 'cbor' | 'hex' | 'leb128' | 'raw';
+}
+
+/**
+ * Pre-configured fields for canister status paths
+ */
+export type Path =
+  | 'time'
+  | 'controllers'
+  | 'subnet'
+  | 'moduleHash'
+  | 'candid'
+  | MetaData
+  | CustomPath;
+
+export type StatusMap = Map<Path | string, Status>;
+
+export type CanisterStatusOptions = {
+  canisterId: Principal;
+  paths?: Path[] | Set<Path>;
+  agentOptions?: HttpAgentOptions;
+  agent?: HttpAgent;
+};
+
+/**
+ *
+ * @param {CanisterStatusOptions} options {@link CanisterStatusOptions}
+ * @param {CanisterStatusOptions['canisterId']} options.canisterId {@link Principal}
+ * @param {CanisterStatusOptions['paths']} options.paths {@link Path[]}
+ * @param {CanisterStatusOptions['agentOptions']} options.agentOptions {@link HttpAgentOptions} optional configuration for the agent used to make the request
+ * @param {CanisterStatusOptions['agent']} options.agent {@link HttpAgent} optional authenticated agent to use to make the canister request. Useful for accessing private metadata under icp:private
+ * @returns {Status} object populated with data from the requested paths
+ * @example
+ * const status = await canisterStatus({
+ *   paths: ['controllers', 'candid'],
+ *   ...options
+ * });
+ *
+ * const controllers = status.get('controllers');
+ */
+export const request = async (options: {
+  canisterId: Principal;
+  paths?: Path[] | Set<Path>;
+  agentOptions?: HttpAgentOptions;
+  agent?: HttpAgent;
+}): Promise<StatusMap> => {
+  const { canisterId, agentOptions, agent, paths } = options;
+
+  const uniquePaths = [...new Set(paths)];
+
+  // Set up agent
+  const effectiveAgent = agent ?? new HttpAgent(agentOptions);
+  if (options.agentOptions?.host && !options.agentOptions.host.startsWith('https://ic0.app')) {
+    await effectiveAgent.fetchRootKey();
+  }
+
+  // Map path options to their correct formats
+  const encodedPaths = uniquePaths.map(path => {
+    return encodePath(path, canisterId);
+  });
+  const status = new Map<string | Path, Status>();
+
+  const promises = uniquePaths.map((path, index) => {
+    return (async () => {
+      try {
+        const response = await effectiveAgent.readState(canisterId, {
+          paths: [encodedPaths[index]],
+        });
+        const cert = new Certificate(response, effectiveAgent);
+        const verified = await cert.verify();
+        if (!verified) {
+          throw new Error(
+            'There was a problem certifying the response data. Please verify your connection to the mainnet, or be sure to call fetchRootKey on your agent if you are developing locally',
+          );
+        }
+
+        const data = cert.lookup(encodePath(uniquePaths[index], canisterId));
+        data;
```

Comment:
is this for debugging?

Reply:
Yes, that was for debugging. I have tooling that makes that be the equivalent of a console.log, but it would be a no-op in actual code. Same goes for the special comment `//?` It should still be removed though